src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -47,73 +82,145 @@ public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
-            // remove it altogether once Roslyn implements workspace/semanticTokens/refresh:
-            // https://github.com/dotnet/roslyn/issues/60441
-            var isFinalized = !context.Document.IsRazorDocument() ||
-                await IsDataFinalizedAsync(project, cancellationToken).ConfigureAwait(false);
+            // If the full compilation is not yet available, we'll try getting a partial one. It may contain inaccurate
+            // results but will speed up how quickly we can respond to the client's request.
+            var document = context.Document.WithFrozenPartialSemantics(cancellationToken);
+            var project = document.Project;
+            var options = _globalOptions.GetClassificationOptions(project.Language) with { ForceFrozenPartialSemanticsForCrossProcessOperations = true };
 
             // The results from the range handler should not be cached since we don't want to cache
             // partial token results. In addition, a range request is only ever called with a whole
             // document request, so caching range results is unnecessary since the whole document
             // handler will cache the results anyway.
             var tokensData = await SemanticTokensHelpers.ComputeSemanticTokensDataAsync(
-                context.Document,
+                document,
                 SemanticTokensHelpers.TokenTypeToIndex,
                 request.Range,
                 options,
                 includeSyntacticClassifications: context.Document.IsRazorDocument(),
                 cancellationToken).ConfigureAwait(false);
 
-            return new RoslynSemanticTokens { Data = tokensData, IsFinalized = isFinalized };
+            // The above call to get semantic tokens may be inaccurate (because we use frozen partial semantics).  Kick
+            // off a request to ensure that the OOP side gets a fully up to compilation for this project.  Once it does
+            // we can optionally choose to notify our caller to do a refresh if we computed a compilation for a new
+            // solution snapshot.
+            if (supportsRefresh)
+                await TryEnqueueRefreshComputationAsync(project, cancellationToken).ConfigureAwait(false);
+
+            return new LSP.SemanticTokens { Data = tokensData };
         }
 
-        private async Task<bool> IsDataFinalizedAsync(Project project, CancellationToken cancellationToken)
+        private async Task TryEnqueueRefreshComputationAsync(Project project, CancellationToken cancellationToken)
         {
-            // We use a combination of IsFullyLoaded + the completed project compilation as the metric
-            // for isFinalized. It may not be completely accurate but this is only a a temporary fix until
-            // workspace/semanticTokens/refresh is implemented.
-            lock (_lock)
+            // Determine the checksum for this project cone.  Note: this should be fast in practice because this is
```

Comment:
question: what's a project "cone"? Is it just another way to say project version?

Reply:
i use "cone" to mean "this project and the projects it depends on".  "cones" are useful for features that need semantics, but couldn't be affected by any projects outside of htat set.  For example if i have projects `A<-B<-C` then to get classification in 'B', i only need the information in 'A' and 'B' as the information in C cannot affect it.  So when we're sync'ing to oop, we actually say "here's the checksum for A+B (but not C), so just sync that information over".